### ISSUE TITLE
Adding `depth` and a vertical RangeError to Arrays

### ISSUE THREAD

**Issue by zfrisch** [tc39/ecma262#1370]
With the increase in functional methods within the ES spec, of which Arrays are most often the target, we can manipulate data in very interesting ways using very little code. One of the more common hindrances to manipulating data is to flatten an array down to one row. 

Thankfully Array.prototype.`flat` was introduced as a proposal and has finally been added to some browsers. 

An interesting thing though is that `flat` has an argument denoting `depth`, (`flat(depth)`) and this is not a readily available property. To obtain it requires iterating over an array and tallying manually, and though this is easier done now than ever, it's still something that might be better addressed as a standard property on the Array prototype. 

I'm not the first person to think this up, and after doing a bit of research I discovered this https://github.com/tc39/proposal-flatMap/issues/68 within the `flat` proposal itself. Although an obvious positive addition, there was a very small problem that was brought up. Self-referencing Arrays. 

`const arr = []; arr.push([arr]);` 

As you can imagine the above creates an Array of infinite depth. I have a hard time wrapping my head around why someone would do this, but it's a fair concern, even though it's most likely a rare circumstance that's (prayerfully) was most likely a mistake. 

The above example as far as I know caused the issue to be closed out and discarded. Based on the merit of a `depth` property, I don't actually believe it should have been. The code given to rebuke it seems an example of something that should _actually_ be corrected within the specification. I think it would be more proper that it would throw an error of some sort, be that a **RangeError: Infinite self-referencing Array** or something like a **TypeError: Invalid Array Assignment**.

Alternatively, and ideally, the **Array.prototype.push** method, before assignment when it creates a list of the pushable items would be a feasible time for a comparison within the items in the list for the same Array reference pointer as is being assigned to. If the Array pointer is found it could then hypothetically automatically transpose that pointer and create a new object with duplicated properties based on the original. Basically Object Assignment:

`const arr = []; arr.push( [ Object.assign([], arr) ]);`

This would completely avoid the infinity array scenario, and even better, would allow for an absolute `depth` which could easily be assigned to a `depth` property.
 
I believe that it's especially worth considering because a standard Array has a horizontal breaking point of 2^32 or 4,294,967,296. This means it has up to 2^32-1 worth of space that is stored within a 32 bit unsigned integer. If we consider also that the horizontal breaking point of an array was created to conserve memory and/or keep performance reasonable, than it is pretty unusual that there is no vertical breaking point for nested arrays either,, and because of that, no `depth`.   

**TLDR;** 
It would be excellent to see a vertical range error at the very least to prevent infinite self-referencing arrays - but even more ideally a change in how Arrays are pushed to, so that an Array _cannot_ reference itself recursively. Either of these would seemingly allow the creation and use of a `depth` property that could be used to simplify a common part of data management.

**Comment by ljharb** [comment]
Thanks for your thoughts! It'd be great if you could submit this to one of the [suggested mediums](https://github.com/tc39/ecma262/blob/master/CONTRIBUTING.md#creating-a-new-proposal) for creating new feature proposals.

**Comment by zfrisch** [comment]
> Thanks for your thoughts! It'd be great if you could submit this to one of the [suggested mediums](https://github.com/tc39/ecma262/blob/master/CONTRIBUTING.md#creating-a-new-proposal) for creating new feature proposals.

Thanks! I will do that as soon as I get a chance!
